### ISSUE TITLE
Add consumers and formalize specification

### ISSUE THREAD

**Issue by pineapplemachine** [pineapplemachine/higher#99]
A consumer shall be a finite state machine. Various functions may call a `push` method which accepts a single arbitrary value as input - typically an element from a sequence. The internal state of any consumer is implementation-dependent; a consumer is defined only by its interface. Consumers likely will have at least these methods:

- `push(value)` to consume a new value
- `done()` when the consumer has reached any termination state, if ever
- `value()` represent termination state: truthy values indicate success, falsey values failure (or pending).
- `copy()` to create a copy of this consumer
- `reset()` to reset the state of this consumer

Consumers are comparable to regular expressions in purpose and concept. In fact, most regular expressions would also be possible to implement as consumers. 

Once, implemented, as separate tasks, consumers should be used:

- By `findFirst`, `findLast`, and `findAll` to find substrings
- By an added `replace` to find substrings
- By `split` to find delimiters
- By an added `match` (name? there is currently another function called "match") to determine if a sequence matches a consumer, i.e. pushing every element to a consumer produces a success state

As part of this task: There shall be default consumer types for comparing to a substring sequence with an optionally specified comparison function, and for matching a series of elements satisfying a predicate function. Eventually, as a separate task, there should also be a default consumer type for acquiring a consumer object from a regular expression.

**Comment by pineapplemachine** [comment]
Consumers should also support reversing where possible

**Comment by pineapplemachine** [comment]
Consumer specification, after lots of rumination. These are attributes of the object to be passed to the `defineConsumer` function.

`summary` - same as sequence types
`docs` - same as sequence types
`tests` - same as sequence types
`converter` - same as sequence `converter` attribute, but applies to consumers instead
`constructor` - same as sequence types
`pushFront` - push an element from the front of a sequence
`pushBack` - push an element from the back of a sequence (optional!)
`pushEnd` - inform the consumer that there are no more elements to push
`done` - return `true` when the consumer either matched or didn't match the subject
`match` - return `true` when the consumer matches the subject
`copy` - return a copy of this consumer

The distinction between `done` and `match` is slightly weird:

`done` is `true` when the consumer won't accept more elements. It means either, "I've matched a complete pattern and if you push any more I will stop matching it", or "I've failed to match the pattern and nothing you push now will change it"

`match` is `true` when the consumer matches the subject. It does not necessarily imply that, after pushing another element, the consumer will no longer match it. To get the full (longest) match of a consumer, one should keep pushing until `done` is true and use the most recent element where `match` was once true to be the end of the match. (If `match` was never true, then there is no match at all.)
